File: app/javascript/channels/chat_room_channel.js
// app/javascript/channels/chat_room_channel.js
import consumer from "./consumer"

document.addEventListener("turbo:load", () => {
  const chatRoomElement = document.getElementById("chat-room");

  if (chatRoomElement) {
    const roomId = chatRoomElement.getAttribute("data-room-id");

    consumer.subscriptions.create({ channel: "ChatRoomChannel", room_id: roomId }, {
      connected() {
        console.log(`Connected to ChatRoom ${roomId}`);
      },

      disconnected() {
        console.log(`Disconnected from ChatRoom ${roomId}`);
      },

      received(data) {
        const messages = document.getElementById("messages");
        messages.insertAdjacentHTML("beforeend", data);
      }
    });
  }
});

File: app/channels/chat_room_channel.rb
class ChatRoomChannel < ApplicationCable::Channel
  def subscribed
    stream_from "chat_room_#{params[:room_id]}"
  end

  def receive(data)
    message = @chat_room.messages.create!(user: current_user, content: data["content"])
    ChatRoomChannel.broadcast_to(@chat_room, {
      content: message.content,
      user_name: message.user.full_name,
      created_at: message.created_at.strftime("%I:%M %p"),
      user_id: message.user.id
    })
  end
end
